src/analyses/useAfterFree.ml

```diff
@@ -0,0 +1,153 @@
+    let cwe_number = if is_double_free then 415 else 416 in
+    let rec offset_might_contain_freed offset =
+      match offset with
+      | NoOffset -> ()
+      | Field (f, o) -> offset_might_contain_freed o
+      | Index (e, o) -> warn_exp_might_contain_freed transfer_fn_name e ctx; offset_might_contain_freed o
+    in
+    let (lval_host, o) = lval in offset_might_contain_freed o; (* Check the lval's offset *)
+    let lval_to_query =
+      match lval_host with
+      | Var _ -> Lval lval
+      | Mem _ -> mkAddrOf lval (* Take the lval's address if its lhost is of the form *p, where p is a ptr *)
+    in
+    match ctx.ask (Queries.MayPointTo lval_to_query) with
+    | a when not (Queries.LS.is_top a) && not (Queries.LS.mem (dummyFunDec.svar, `NoOffset) a) ->
+      begin try
+          let v, o = Queries.LS.choose a in
+          if ctx.ask (Queries.IsHeapVar v) && D.mem v state then
+            M.warn ~category:(Behavior undefined_behavior) ~tags:[CWE cwe_number] "lval (%s) in \"%s\" points to a maybe freed memory region" v.vname transfer_fn_name
+        with Not_found -> ()
+      end
+    | _ -> ()
+
+  and warn_exp_might_contain_freed ?(is_double_free = false) (transfer_fn_name:string) (exp:exp) ctx =
+    match exp with
+    (* Base recursion cases *)
+    | Const _
+    | SizeOf _
+    | SizeOfStr _
+    | AlignOf _
+    | AddrOfLabel _ -> ()
+    (* Non-base cases *)
+    | Real e
+    | Imag e
+    | SizeOfE e
+    | AlignOfE e
+    | UnOp (_, e, _)
+    | CastE (_, e) -> warn_exp_might_contain_freed ~is_double_free transfer_fn_name e ctx
+    | BinOp (_, e1, e2, _) ->
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e1 ctx;
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e2 ctx
+    | Question (e1, e2, e3, _) ->
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e1 ctx;
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e2 ctx;
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e3 ctx
+    (* Lval cases (need [warn_lval_might_contain_freed] for them) *)
+    | Lval lval
+    | StartOf lval
+    | AddrOf lval -> warn_lval_might_contain_freed ~is_double_free transfer_fn_name lval ctx
+
+
+  (* TRANSFER FUNCTIONS *)
+
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    warn_lval_might_contain_freed "assign" lval ctx;
+    warn_exp_might_contain_freed "assign" rval ctx;
+    ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    warn_exp_might_contain_freed "branch" exp ctx;
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    Option.iter (fun x -> warn_exp_might_contain_freed "return" x ctx) exp;
+    ctx.local
+
+  let enter ctx (lval:lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    let caller_state = ctx.local in
+    Option.iter (fun x -> warn_lval_might_contain_freed "enter" x ctx) lval;
```

Comment:
I think this check should be made in `combine_assign`: Consider a case where the calle calls `free()` for something that `lval` may point to.

Reply:
Done in https://github.com/goblint/analyzer/pull/1050/commits/d518665e4d8c0cba299f4d5a34c8a3bf4c1aaa67